File: ios/runtime/XTRFont.m
//
//  XTRFont.m
//  XTSample
//
//  Created by 崔明辉 on 2017/8/30.
//  Copyright © 2017年 UED Center, YY Inc. All rights reserved.
//

#import "XTRFont.h"
#import "XTRContext.h"
#import <XT-Mem/XTMemoryManager.h>

@implementation XTRFont

+ (NSString *)create:(JSValue *)pointSize
         fontWeight:(JSValue *)fontWeight
          fontStyle:(JSValue *)fontStyle
         familyName:(JSValue *)familyName {
    XTRFont *obj = [XTRFont new];
    if (familyName != nil && [familyName toString].length > 0) {
        obj.innerObject = [UIFont fontWithName:[familyName toString] size:[pointSize toDouble]];
    }
    else if ([[fontStyle toString] isEqualToString:@"italic"]) {
        obj.innerObject = [UIFont italicSystemFontOfSize:[pointSize toDouble]];
    }
    else {
        if ([UIDevice currentDevice].systemVersion.floatValue >= 8.2) {
            obj.innerObject = [UIFont systemFontOfSize:[pointSize toDouble] weight:[fontWeight toDouble]];
        }
        else {
            if ([fontWeight toDouble] >= 700) {
                obj.innerObject = [UIFont boldSystemFontOfSize:[pointSize toDouble]];
            }
            else {
                obj.innerObject = [UIFont systemFontOfSize:[pointSize toDouble]];
            }
        }
    }
    XTManagedObject *managedObject = [[XTManagedObject alloc] initWithObject:obj];
    obj.objectUUID = managedObject.objectUUID;
    [XTMemoryManager add:managedObject];
    return managedObject.objectUUID;
}

+ (NSString *)create:(UIFont *)font {
    XTRFont *obj = [XTRFont new];
    obj.innerObject = font;
    XTManagedObject *managedObject = [[XTManagedObject alloc] initWithObject:obj];
    obj.objectUUID = managedObject.objectUUID;
    [XTMemoryManager add:managedObject];
    return managedObject.objectUUID;
}

+ (NSString *)name {
    return @"XTRFont";
}

- (void)dealloc {
#ifdef LOGDEALLOC
    NSLog(@"XTRFont dealloc.");
#endif
}

+ (NSString *)xtr_familyName:(NSString *)objectRef {
    XTRFont *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[XTRFont class]]) {
        return obj.innerObject.familyName;
    }
    return nil;
}

+ (NSNumber *)xtr_pointSize:(NSString *)objectRef {
    XTRFont *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[XTRFont class]]) {
        return @(obj.innerObject.pointSize);
    }
    return nil;
}

+ (NSString *)xtr_fontWeight:(NSString *)objectRef {
    XTRFont *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[XTRFont class]]) {
        NSString *UIUsageAttribute = obj.innerObject.fontDescriptor.fontAttributes[@"NSCTFontUIUsageAttribute"];
        if ([UIUsageAttribute isEqualToString:@"CTFontBlackUsage"]) {
            return @"700";
        }
        return @"400";
    }
    return nil;
}

+ (NSString *)xtr_fontStyle:(NSString *)objectRef {
    XTRFont *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[XTRFont class]]) {
        NSString *UIUsageAttribute = obj.innerObject.fontDescriptor.fontAttributes[@"NSCTFontUIUsageAttribute"];
        if ([UIUsageAttribute isEqualToString:@"CTFontObliqueUsage"]) {
            return @"italic";
        }
        return @"normal";
    }
    return nil;
}

@end

File: android/xtruntime/src/main/java/com/opensource/xtruntime/XTRFont.kt
package com.opensource.xtruntime

import android.graphics.Typeface
import com.eclipsesource.v8.V8Object
import com.opensource.xtmem.XTManagedObject
import com.opensource.xtmem.XTMemoryManager

/**
 * Created by cuiminghui on 2018/1/9.
 */

class XTRFont(val pointSize: Double, val fontWeight: String, val fontStyle: String, val familyName: String): XTRComponentInstance {

    override var objectUUID: String? = null

    val typeface: Typeface

    init {
        var typefaceStyle = Typeface.NORMAL
        if (fontWeight == "700" && fontStyle == "italic") {
            typefaceStyle = Typeface.BOLD_ITALIC
        }
        else if (fontWeight == "700") {
            typefaceStyle = Typeface.BOLD
        }
        else if (fontStyle == "italic") {
            typefaceStyle = Typeface.ITALIC
        }
        typeface = Typeface.create(familyName, typefaceStyle)
    }

    class JSExports(val context: XTRContext): XTRComponentExport() {

        override val name: String = "XTRFont"

        override fun exports(): V8Object {
            val exports = V8Object(context.runtime)
            exports.registerJavaMethod(this, "create", "create", arrayOf(Double::class.java, String::class.java, String::class.java, String::class.java))
            exports.registerJavaMethod(this, "xtr_familyName", "xtr_familyName", arrayOf(String::class.java))
            exports.registerJavaMethod(this, "xtr_pointSize", "xtr_pointSize", arrayOf(String::class.java))
            exports.registerJavaMethod(this, "xtr_fontWeight", "xtr_fontWeight", arrayOf(String::class.java))
            exports.registerJavaMethod(this, "xtr_fontStyle", "xtr_fontStyle", arrayOf(String::class.java))
            return exports
        }

        fun xtr_familyName(objectRef: String): String? {
            return (XTMemoryManager.find(objectRef) as? XTRFont)?.familyName
        }

        fun xtr_pointSize(objectRef: String): Double {
            return (XTMemoryManager.find(objectRef) as? XTRFont)?.pointSize ?: 0.0
        }

        fun xtr_fontWeight(objectRef: String): String? {
            return (XTMemoryManager.find(objectRef) as? XTRFont)?.fontWeight ?: "400"
        }

        fun xtr_fontStyle(objectRef: String): String? {
            return (XTMemoryManager.find(objectRef) as? XTRFont)?.fontStyle ?: "normal"
        }

        fun create(pointSize: Double, fontWeight: String, fontStyle: String, familyName: String): String {
            val font = XTRFont(pointSize, fontWeight, fontStyle, familyName)
            val managedObject = XTManagedObject(font)
            font.objectUUID = managedObject.objectUUID
            XTMemoryManager.add(managedObject)
            return managedObject.objectUUID
        }

    }

}